hal/ia32/console.c

```diff
@@ -21,133 +21,343 @@
 #include "ia32.h"
 
 
-#define VRAM_MONO     (void *)0xb0000
-#define VRAM_COLOR    (void *)0xb8000
+/* ANSI escape sequence states */
+enum {
+	esc_init, /* normal */
+	esc_esc,  /* esc */
+	esc_csi,  /* esc [ */
+	esc_csiqm /* esc [ ? */
+};
 
-#define BASE_MONO     0x3b4  /* crtc index register address mono */
-#define BASE_COLOR    0x3d4  /* crtc index register address color */
 
-#define CRTC_CURSORH  0x0e   /* cursor address mid */
-#define CRTC_CURSORL  0x0f   /* cursor address low */
+/* ANSI code to VGA foreground character attribute conversion table */
+static const unsigned char ansi2fg[] = {
+	0x00, 0x04, 0x02, 0x06,
+	0x01, 0x05, 0x03, 0x07
+};
 
-#define MAIN_MISCIN   0x3cc  /* miscellaneous input register */
+
+/* ANSI code to VGA background character attribute conversion table */
+static const unsigned char ansi2bg[] = {
+	0x00, 0x40, 0x20, 0x60,
+	0x10, 0x50, 0x30, 0x70
+};
 
 
 struct {
+	volatile u16 *vram;      /* Video memory */
+	void *crtc;              /* CRT controller register */
+	unsigned int rows;       /* Console height */
+	unsigned int cols;       /* Console width */
+	unsigned char attr;      /* Character attribute */
+	unsigned char esc;       /* Escape sequence state */
+	unsigned char parmi;     /* Escape sequence parameter index */
+	unsigned char parms[10]; /* Escape sequence parameters buffer */
 	spinlock_t spinlock;
-	unsigned char rows;
-	unsigned char maxcol;
-	unsigned char row;
-	unsigned char col;
-	u16 *vram;
-	void *crtc;
+} halconsole_common;
+
+
+static void console_memset(volatile u16 *vram, u16 val, unsigned int n)
+{
+	unsigned int i;
 
-	int attr;
-} console;
+	for (i = 0; i < n; i++)
+		*(vram + i) = val;
```

Comment:
Maybe inline assembler and using `movs`. It will be faster.

Reply:
This code is derived from [plo](https://github.com/phoenix-rtos/plo/blob/f5401788f72d58658a2aa1879ccfabc475cc900b/hal/ia32/console-vga.c#L79-L314) so in this PR I would not like to add significant modifications.